### PR TITLE
Feature/arangosearch-maintenance-rework

### PR DESCRIPTION
**Pull request by gnusi**
### Scope & Purpose

*(Please describe the changes in this PR for reviewers - **mandatory**)*

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [x] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [x] Backports required for: *3.7*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket number: ES-662
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12966/

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [x] Added entry to *Release Notes*
https://github.com/arangodb/docs/pull/593
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/IResearch/IResearchLink.cpp [hunk 1]
```diff
@@ -253,42 +254,351 @@ bool readTick(irs::bytes_ref const& payload, TRI_voc_tick_t& tick) noexcept {
...
+
+  auto linkLock = irs::make_unique_lock(link->mutex(), std::try_to_lock);
+
+  if (!linkLock.owns_lock()) {
+    LOG_TOPIC("eb0de", DEBUG, iresearch::TOPIC)
+        << "failed to acquire the lock while committing the link '" << id
+        << "', runId '" << size_t(&runId) << "'";
+
+    // blindly reschedule commit task
+    schedule(commitIntervalMsec);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp:363]: added corresponding check in `Task::schedule(...)`

File: arangod/IResearch/IResearchLink.cpp [hunk 2]
```diff
@@ -253,42 +254,351 @@ bool readTick(irs::bytes_ref const& payload, TRI_voc_tick_t& tick) noexcept {
...
+      ++state->pendingCommits;
+      schedule(commitIntervalMsec);
+
+      if (code == IResearchLink::CommitResult::DONE) {
+        state->noopCommitCount = 0;
+        state->noopConsolidationCount = 0;
+
+        if (state->pendingConsolidations < MAX_PENDING_CONSOLIDATIONS &&
+            ++state->nonEmptyCommits > MAX_NON_EMPTY_COMMITS) {
+          link->scheduleConsolidation(consolidationIntervalMsec);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp:412]: added corresponding check in `Task::schedule(...)`

File: arangod/IResearch/IResearchLink.cpp [hunk 3]
```diff
@@ -253,42 +254,351 @@ bool readTick(irs::bytes_ref const& payload, TRI_voc_tick_t& tick) noexcept {
...
+          state->nonEmptyCommits = 0;
+        }
+      }
+    } else {
+      state->nonEmptyCommits = 0;
+      ++state->noopCommitCount;
+
+      for (auto count = state->pendingCommits.load(); count < 1; ) {
+        if (state->pendingCommits.compare_exchange_strong(count, count + 1)) {
+          schedule(commitIntervalMsec);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp:422]: added corresponding check in `Task::schedule(...)`

File: arangod/IResearch/IResearchLink.cpp [hunk 4]
```diff
@@ -253,42 +254,351 @@ bool readTick(irs::bytes_ref const& payload, TRI_voc_tick_t& tick) noexcept {
...
+
+  IResearchLink::CommitResult code = IResearchLink::CommitResult::UNDEFINED;
+
+  auto reschedule = scopeGuard([&code, link, this](){
+    constexpr size_t MAX_NON_EMPTY_COMMITS = 10;
+    constexpr size_t MAX_PENDING_CONSOLIDATIONS = 3;
+
+    if (code != IResearchLink::CommitResult::NO_CHANGES) {
+      ++state->pendingCommits;
+      schedule(commitIntervalMsec);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp:404]: added corresponding check in `Task::schedule(...)`

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-linux/14451/

**Review by Dronplane — APPROVED** [review]
LGTM

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12941/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12942/

**Comment by gnusi** [comment]
mac build
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-mac-onlycatch/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12959/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12961/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12963/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12966/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12971/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/

**Pull request merged**